src/target/xtensa/xtensa_fileio.c

```diff
@@ -0,0 +1,201 @@
+ *                                                                         *
+ *   This program is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
+ *   GNU General Public License for more details.                          *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public License     *
+ *   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
+ ***************************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "xtensa_chip.h"
+#include "xtensa_fileio.h"
+#include "xtensa.h"
+
+#define XTENSA_SYSCALL(x)			XT_INS_BREAK(x,1,14)
+#define XTENSA_SYSCALL_SZ			3
+#define XTENSA_SYSCALL_LEN_MAX		255
+
+
+int xtensa_fileio_init(struct target *target)
+{
+	target->fileio_info = malloc(sizeof(struct gdb_fileio_info));
+	target->fileio_info->identifier = malloc(XTENSA_SYSCALL_LEN_MAX + 1);
+	return ERROR_OK;
+}
+
+/**
+ * Checks for and processes an Xtensa File-IO request.  
+ *
+ * Return ERROR_OK if request was found and handled; or 
+ * return ERROR_FAIL if no request was detected.
+ */
+int xtensa_fileio_detect_proc(struct target *target)
+{
+	struct xtensa *xtensa = target_to_xtensa(target);
+	int retval;
+
+	uint32_t dbg_cause = xtensa_cause_get(target);
+	if (((dbg_cause & (DEBUGCAUSE_BI|DEBUGCAUSE_BN)) == 0) || xtensa->halt_request) {
+		return ERROR_FAIL;
+	}
+
+	uint8_t brk_insn_buf[sizeof(uint32_t)] = {0};
+	uint32_t pc = xtensa_reg_get(target, XT_REG_IDX_PC);
+	retval = target_read_memory(target,
+		pc,
+		XTENSA_SYSCALL_SZ,
+		1,
+		(uint8_t *)brk_insn_buf);
+	if (retval != ERROR_OK) {
+		LOG_ERROR("Failed to read break instruction!");
+		return ERROR_FAIL;
+	}
+	if (buf_get_u32(brk_insn_buf, 0, 32) != XTENSA_SYSCALL(xtensa)) {
+		return ERROR_FAIL;
+	}
+
+	LOG_DEBUG("File-I/O: syscall breakpoint found at 0x%x", pc);
+	xtensa->proc_syscall = true;
+	return ERROR_OK;
+}
+
+int xtensa_get_gdb_fileio_info(struct target *target, struct gdb_fileio_info *fileio_info)
+{
+	/* fill syscall parameters to file-I/O info */
+	if (!fileio_info) {
+		LOG_ERROR("File-I/O data structure uninitialized");
+		return ERROR_FAIL;
+	}
+
+	struct xtensa *xtensa = target_to_xtensa(target);
+	if (!xtensa->proc_syscall) {
+		return ERROR_FAIL;
+	}
+
+	uint32_t syscall = xtensa_reg_get(target, XTENSA_SYSCALL_OP_REG);
+	uint32_t arg0 = xtensa_reg_get(target, XT_REG_IDX_A6);
+	uint32_t arg1 = xtensa_reg_get(target, XT_REG_IDX_A3);
+	uint32_t arg2 = xtensa_reg_get(target, XT_REG_IDX_A4);
+	uint32_t arg3 = xtensa_reg_get(target, XT_REG_IDX_A5);
+	int retval = ERROR_OK;
+
+	LOG_DEBUG("File-I/O: syscall 0x%x 0x%x 0x%x 0x%x 0x%x", syscall, arg0, arg1, arg2, arg3);
+
+	switch (syscall) {
+	case XTENSA_SYSCALL_OPEN:
```

Comment:
in the semihosting_common.c, these syscalls are handled. Can't we use them?

Reply:
Will address this later.  I'd like to merge semihosting/fileio, but the semihosting implementation seems more arm-centric.